File: 0509-fibonacci-number/0509-fibonacci-number.cpp
class Solution {
private:
    int fibn(int n, vector<int> & dp){
        if(n==0)
         return 0;
        if(n==1)
         return 1;
        if(dp[n]!=-1) return dp[n];
        return dp[n] = fibn(n-2,dp) + fibn(n-1,dp);
    }
public:
    int fib(int n) {

        vector<int> dp(n+1,-1); 
        return fibn(n,dp);
    }
};

File: 0509-fibonacci-number/0509-fibonacci-number.rs
impl Solution {
    fn fibonacci(n:i32, dp: &mut Vec<i32>)->i32{
        if n==1 {
            return 1;
        }
        if n==0 {
            return 0;
        }
        if dp[n as usize] != -1 {
            return dp[n as usize];
        }
        dp[n as usize] = Solution::fibonacci(n-1,dp) + Solution::fibonacci(n-2,dp);
        return dp[n as usize];
    }
    pub fn fib(n: i32) -> i32 {
        let mut dp : Vec<i32> = vec![-1;(n+1) as usize];
        Solution::fibonacci(n,&mut dp)
    }
}